File: tobas_setup_assistant/src/tobas_setup_assistant/setting_widgets/custom_joints.py
from __future__ import annotations
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from ..setup_assistant import SetupAssistant

from enum import Enum
from overrides import override
from typing import List
from PyQt5.QtWidgets import QLabel

from tobas_std_tools_py.string import title_from_snake
from tobas_rqt_tools.widgets import DoubleSpinBox, ComboBox, TableWidget
from tobas_rqt_tools.messages import q_error_named
from tobas_kdl_sympy.joint import HardwareInterface, JointType

from .base_setting import BaseSettingWidget


class CustomJointField(Enum):
    JOINT_NAME = 0
    HOME_POSITION = 1
    MIN_POSITION = 2
    MAX_POSITION = 3
    COMMAND_TYPE = 4
    P_GAIN = 5
    I_GAIN = 6
    D_GAIN = 7


class CustomJointsWidget(BaseSettingWidget):
    NAME = "Custom Joints"

    POSITION = "position"
    VELOCITY = "velocity"
    EFFORT = "effort"

    DEFAULT_P_GAIN = 10.0
    DEFAULT_I_GAIN = 0.1
    DEFAULT_D_GAIN = 1.0

    COL_WIDTH = 120
    POS_DECIMALS = 3
    GAIN_DECIMALS = 3

    def __init__(self, main: SetupAssistant) -> None:
        title_text = "Define Custom Joints"
        abst_text = (
            "Configure the settings for joints with transmissions "
            "other than those in the propulsion system and fixed-wing control surfaces."
        )
        super().__init__(main, title_text, abst_text)

        self._table = TableWidget(0, len(CustomJointField))
        self._table.setHorizontalHeaderLabels([title_from_snake(item.name) for item in CustomJointField])
        for c in range(self._table.columnCount()):
            self._table.setColumnWidth(c, self.COL_WIDTH)
        self._rows.addWidget(self._table)

        self._rows.addStretch()

    @override
    def update_internal_data_structures(self) -> None:
        self._table.remove_all()
        row = 0

        for joint in self._main.urdf_parser.get_joints():
            # ジョイントタイプが回転または直動でない場合はスキップ
            if joint.type not in {JointType.REVOLUTE, JointType.CONTINUOUS, JointType.PRISMATIC}:
                continue

            # トランスミッションを持たない場合はスキップ
            hi = self._main.urdf_parser.hardware_interface(joint.name)
            if hi is None:
                continue

            self._table.insertRow(row)

            jnt_name = QLabel(joint.name)

            home_pos = DoubleSpinBox()
            min_pos = DoubleSpinBox()
            max_pos = DoubleSpinBox()

            home_pos.setDecimals(self.POS_DECIMALS)
            min_pos.setDecimals(self.POS_DECIMALS)
            max_pos.setDecimals(self.POS_DECIMALS)

            home_pos.setValue(0.0)
            if joint.limit is not None:
                home_pos.setMinimum(joint.limit.lower)
                home_pos.setMaximum(joint.limit.upper)
                min_pos.setMinimum(joint.limit.lower)
                min_pos.setMaximum(joint.limit.upper)
                min_pos.setValue(joint.limit.lower)
                max_pos.setMinimum(joint.limit.lower)
                max_pos.setMaximum(joint.limit.upper)
                max_pos.setValue(joint.limit.upper)

            if joint.type == JointType.REVOLUTE or joint.type == JointType.CONTINUOUS:
                home_pos.setSuffix(" rad")
                min_pos.setSuffix(" rad")
                max_pos.setSuffix(" rad")
            elif joint.type == JointType.PRISMATIC:
                home_pos.setSuffix(" m")
                min_pos.setSuffix(" m")
                max_pos.setSuffix(" m")
            else:
                raise RuntimeError(f"Joint type '{joint.type}' is unexpected.")

            cmd_type = ComboBox()
            p_gain = DoubleSpinBox()
            i_gain = DoubleSpinBox()
            d_gain = DoubleSpinBox()

            p_gain.setDecimals(self.GAIN_DECIMALS)
            i_gain.setDecimals(self.GAIN_DECIMALS)
            d_gain.setDecimals(self.GAIN_DECIMALS)

            p_gain.setValue(self.DEFAULT_P_GAIN)
            i_gain.setValue(self.DEFAULT_I_GAIN)
            d_gain.setValue(self.DEFAULT_D_GAIN)

            if hi == HardwareInterface.POSITION:
                cmd_type.addItem(self.POSITION)
                p_gain.setEnabled(False)
                i_gain.setEnabled(False)
                d_gain.setEnabled(False)
            elif hi == HardwareInterface.VELOCITY:
                cmd_type.addItem(self.POSITION)
                cmd_type.addItem(self.VELOCITY)
                cmd_type.setCurrentText(self.VELOCITY)
            elif hi == HardwareInterface.EFFORT:
                cmd_type.addItem(self.POSITION)
                cmd_type.addItem(self.VELOCITY)
                cmd_type.addItem(self.EFFORT)
                cmd_type.setCurrentText(self.EFFORT)
            else:
                raise RuntimeError(f"Unknown hardware interface: {hi.value}")

            self._table.setCellWidget(row, CustomJointField.JOINT_NAME.value, jnt_name)
            self._table.setCellWidget(row, CustomJointField.HOME_POSITION.value, home_pos)
            self._table.setCellWidget(row, CustomJointField.MIN_POSITION.value, min_pos)
            self._table.setCellWidget(row, CustomJointField.MAX_POSITION.value, max_pos)
            self._table.setCellWidget(row, CustomJointField.COMMAND_TYPE.value, cmd_type)
            self._table.setCellWidget(row, CustomJointField.P_GAIN.value, p_gain)
            self._table.setCellWidget(row, CustomJointField.I_GAIN.value, i_gain)
            self._table.setCellWidget(row, CustomJointField.D_GAIN.value, d_gain)

            row += 1

    @override
    def is_valid(self) -> bool:
        for i in range(self.count()):
            jnt_name = self.get_joint_name(i)
            home_pos = self.get_home_position(i)
            min_pos = self.get_min_position(i)
            max_pos = self.get_max_position(i)
            if min_pos > max_pos:
                q_error_named(self, self.NAME, f'Position limit of joint "{jnt_name}" is invalid.')
                return False
            if home_pos < min_pos or max_pos < home_pos:
                q_error_named(self, self.NAME, f'Home position of joint "{jnt_name}" is out of limit.')
                return False

        return True

    @override
    def dump_settings(self) -> dict:
        res = dict()

        for row in range(self.count()):
            res[self.get_joint_name(row)] = {
                CustomJointField.HOME_POSITION.name: self.get_home_position(row),
                CustomJointField.MIN_POSITION.name: self.get_min_position(row),
                CustomJointField.MAX_POSITION.name: self.get_max_position(row),
                CustomJointField.COMMAND_TYPE.name: self.get_command_type(row),
                CustomJointField.P_GAIN.name: self.get_p_gain(row),
                CustomJointField.I_GAIN.name: self.get_i_gain(row),
                CustomJointField.D_GAIN.name: self.get_d_gain(row),
            }

        return res

    @override
    def load_settings(self, data: dict) -> None:
        for joint_name, data_ in data.items():
            row = self._get_row(joint_name)
            if row < 0:
                q_error_named(self, self.NAME, f'"{joint_name}" does not exist in the custom joint list.')
                continue

            self.set_home_position(row, data_[CustomJointField.HOME_POSITION.name])
            self.set_min_position(row, data_[CustomJointField.MIN_POSITION.name])
            self.set_max_position(row, data_[CustomJointField.MAX_POSITION.name])
            self.set_command_type(row, data_[CustomJointField.COMMAND_TYPE.name])
            self.set_p_gain(row, data_[CustomJointField.P_GAIN.name])
            self.set_i_gain(row, data_[CustomJointField.I_GAIN.name])
            self.set_d_gain(row, data_[CustomJointField.D_GAIN.name])

    def count(self) -> int:
        """ジョイント数．"""
        return self._table.rowCount()

    def get_joint_name(self, row: int) -> str:
        cell: QLabel = self._table.cellWidget(row, CustomJointField.JOINT_NAME.value)
        return cell.text()

    def set_joint_name(self, row: int, joint_name) -> None:
        cell: QLabel = self._table.cellWidget(row, CustomJointField.JOINT_NAME.value)
        cell.setText(joint_name)

    def get_home_position(self, row: int) -> float:
        cell: DoubleSpinBox = self._table.cellWidget(row, CustomJointField.HOME_POSITION.value)
        return cell.value()

    def set_home_position(self, row: int, value: float) -> None:
        cell: DoubleSpinBox = self._table.cellWidget(row, CustomJointField.HOME_POSITION.value)
        cell.setValue(value)

    def get_min_position(self, row: int) -> float:
        cell: DoubleSpinBox = self._table.cellWidget(row, CustomJointField.MIN_POSITION.value)
        return cell.value()

    def set_min_position(self, row: int, value: float) -> None:
        cell: DoubleSpinBox = self._table.cellWidget(row, CustomJointField.MIN_POSITION.value)
        cell.setValue(value)

    def get_max_position(self, row: int) -> float:
        cell: DoubleSpinBox = self._table.cellWidget(row, CustomJointField.MAX_POSITION.value)
        return cell.value()

    def set_max_position(self, row: int, value: float) -> None:
        cell: DoubleSpinBox = self._table.cellWidget(row, CustomJointField.MAX_POSITION.value)
        cell.setValue(value)

    def get_command_type(self, row: int) -> str:
        cell: ComboBox = self._table.cellWidget(row, CustomJointField.COMMAND_TYPE.value)
        return cell.currentText()

    def set_command_type(self, row: int, command_type: str) -> None:
        cell: ComboBox = self._table.cellWidget(row, CustomJointField.COMMAND_TYPE.value)
        cell.setCurrentText(command_type)

    def get_p_gain(self, row: int) -> float:
        cell: DoubleSpinBox = self._table.cellWidget(row, CustomJointField.P_GAIN.value)
        return cell.value()

    def set_p_gain(self, row: int, value: float) -> None:
        cell: DoubleSpinBox = self._table.cellWidget(row, CustomJointField.P_GAIN.value)
        cell.setValue(value)

    def get_i_gain(self, row: int) -> float:
        cell: DoubleSpinBox = self._table.cellWidget(row, CustomJointField.I_GAIN.value)
        return cell.value()

    def set_i_gain(self, row: int, value: float) -> None:
        cell: DoubleSpinBox = self._table.cellWidget(row, CustomJointField.I_GAIN.value)
        cell.setValue(value)

    def get_d_gain(self, row: int) -> float:
        cell: DoubleSpinBox = self._table.cellWidget(row, CustomJointField.D_GAIN.value)
        return cell.value()

    def set_d_gain(self, row: int, value: float) -> None:
        cell: DoubleSpinBox = self._table.cellWidget(row, CustomJointField.D_GAIN.value)
        cell.setValue(value)

    def get_joint_names(self) -> List[str]:
        return [self.get_joint_name(row) for row in range(self.count())]

    def get_controller_type(self, row: int) -> str:
        jnt_name = self.get_joint_name(row)
        hi = self._main.urdf_parser.hardware_interface(jnt_name)
        if hi == HardwareInterface.POSITION:
            group = "position_controllers"
        elif hi == HardwareInterface.VELOCITY:
            group = "velocity_controllers"
        elif hi == HardwareInterface.EFFORT:
            group = "effort_controllers"
        else:
            raise RuntimeError()

        cmd_type = self.get_command_type(row)
        if cmd_type == self.POSITION:
            controller = "JointPositionController"
        elif cmd_type == self.VELOCITY:
            controller = "JointVelocityController"
        elif cmd_type == self.EFFORT:
            controller = "JointEffortController"
        else:
            raise RuntimeError()

        return f"{group}/{controller}"

    def pid_enabled(self, row: int) -> bool:
        jnt_name = self.get_joint_name(row)
        hi = self._main.urdf_parser.hardware_interface(jnt_name)
        cmd_type: str = self.get_command_type(row)

        if hi == HardwareInterface.POSITION and cmd_type == self.POSITION:
            return False
        if hi == HardwareInterface.VELOCITY and cmd_type == self.VELOCITY:
            return False
        if hi == HardwareInterface.EFFORT and cmd_type == self.EFFORT:
            return False
        return True

    def _get_row(self, joint_name: str) -> int:
        for row in range(self.count()):
            if self.get_joint_name(row) == joint_name:
                return row
        else:
            return -1

File: tobas_gui/tobas_setup_assistant/src/setting_widgets/custom_joints.cpp
#include <tobas_yaml_tools/convert/qstring.hpp>
#include <tobas_qt_tools/widgets/spin_box.hpp>
#include <tobas_qt_tools/widgets/combo_box.hpp>
#include <tobas_qt_tools/message.hpp>

#include "tobas_setup_assistant/setting_tabs/custom_joints.hpp"

namespace gui
{
namespace setup_assistant
{
CustomJointsWidget::CustomJointsWidget(const RobotInfo& robot) : robot_(robot)
{
}

const char* CustomJointsWidget::name() const
{
  return "Custom Joints";
}

const char* CustomJointsWidget::title() const
{
  return "Define Custom Joints";
}

const char* CustomJointsWidget::description() const
{
  return "Configure the settings for joints with transmissions "
         "other than those in the propulsion system and fixed-wing control surfaces.";
}

void CustomJointsWidget::onInit()
{
  const QStringList labels{ kLinkNameLabel, kJointNameLabel, kHomePosLabel, kMinPosLabel, kMaxPosLabel,
                            kCmdTypeLabel,  kPGainLabel,     kIGainLabel,   kDGainLabel };
  table_ = new qt::TableWidget(0, labels.size());
  table_->setHorizontalHeaderLabels(labels);
  for (int c = 0; c < table_->columnCount(); ++c)
    table_->setColumnWidth(c, kColWidth);
  addWidget(table_);
}

void CustomJointsWidget::onOpened()
{
  return;
}

void CustomJointsWidget::updateInternalDataStructures()
{
  table_->removeAll();
  int row = 0;

  for (const auto& [link_name, elem] : robot_.tree().getSegments())
  {
    const auto& joint = elem.segment.joint();

    // 可動ジョイントでない場合はスキップ
    if (joint.type == kdl::Joint::Fixed)
      continue;

    // トランスミッションが無効な場合はスキップ
    const auto hi = robot_.hardwareInterface(joint.name);
    if (hi < 0)
      return;

    table_->insertRow(row);

    const auto link_name_label = new QLabel(QString::fromStdString(link_name));
    const auto jnt_name_label = new QLabel(QString::fromStdString(joint.name));

    const auto home_pos = new qt::DoubleSpinBox();
    const auto min_pos = new qt::DoubleSpinBox();
    const auto max_pos = new qt::DoubleSpinBox();

    home_pos->setDecimals(kPosDecimals);
    min_pos->setDecimals(kPosDecimals);
    max_pos->setDecimals(kPosDecimals);

    home_pos->setMinimum(joint.lower_limit);
    min_pos->setMinimum(joint.lower_limit);
    max_pos->setMinimum(joint.lower_limit);

    home_pos->setMaximum(joint.upper_limit);
    min_pos->setMaximum(joint.upper_limit);
    max_pos->setMaximum(joint.upper_limit);

    home_pos->setValue(0.);
    min_pos->setValue(joint.lower_limit);
    max_pos->setValue(joint.upper_limit);

    switch (joint.type)
    {
      case kdl::Joint::RotAxis:
        home_pos->setSuffix(" rad");
        min_pos->setSuffix(" rad");
        max_pos->setSuffix(" rad");
        break;
      case kdl::Joint::TransAxis:
        home_pos->setSuffix(" m");
        min_pos->setSuffix(" m");
        max_pos->setSuffix(" m");
        break;
      default:
        qt::qErrorBox(this, QString::fromStdString(std::format("Unexpected joint type: {}", (int)joint.type)));
        continue;
    }

    const auto cmd_type = new qt::ComboBox();

    const auto p_gain = new qt::DoubleSpinBox();
    const auto i_gain = new qt::DoubleSpinBox();
    const auto d_gain = new qt::DoubleSpinBox();

    p_gain->setDecimals(kGainDecimals);
    i_gain->setDecimals(kGainDecimals);
    d_gain->setDecimals(kGainDecimals);

    p_gain->setValue(kDefaultPGain);
    i_gain->setValue(kDefaultIGain);
    d_gain->setValue(kDefaultDGain);

    switch (hi)
    {
      case hw_interface::POSITION:
        cmd_type->addItem(kPositionLabel);
        p_gain->setEnabled(false);
        i_gain->setEnabled(false);
        d_gain->setEnabled(false);
        break;
      case hw_interface::VELOCITY:
        cmd_type->addItem(kPositionLabel);
        cmd_type->addItem(kVelocityLabel);
        cmd_type->setCurrentText(kVelocityLabel);
        break;
      case hw_interface::EFFORT:
        cmd_type->addItem(kPositionLabel);
        cmd_type->addItem(kVelocityLabel);
        cmd_type->addItem(kEffortLabel);
        cmd_type->setCurrentText(kEffortLabel);
        break;
      default:
        qt::qErrorBox(this, QString::fromStdString(std::format("Unknown hardware interface: {}", (int)hi)));
        continue;
    }

    table_->setCellWidget(row, static_cast<int>(LINK_NAME), link_name_label);
    table_->setCellWidget(row, static_cast<int>(JOINT_NAME), jnt_name_label);
    table_->setCellWidget(row, static_cast<int>(HOME_POSITION), home_pos);
    table_->setCellWidget(row, static_cast<int>(MIN_POSITION), min_pos);
    table_->setCellWidget(row, static_cast<int>(MAX_POSITION), max_pos);
    table_->setCellWidget(row, static_cast<int>(COMMAND_TYPE), cmd_type);
    table_->setCellWidget(row, static_cast<int>(P_GAIN), p_gain);
    table_->setCellWidget(row, static_cast<int>(I_GAIN), i_gain);
    table_->setCellWidget(row, static_cast<int>(D_GAIN), d_gain);

    ++row;
  }
}

bool CustomJointsWidget::isValid()
{
  for (int row = 0; row < count(); ++row)
  {
    const auto jnt_name = getJointName(row);
    const auto home_pos = getHomePosition(row);
    const auto min_pos = getMinPosition(row);
    const auto max_pos = getMaxPosition(row);

    if (min_pos > max_pos)
    {
      qt::qErrorBox(this, "Position limit of joint \"" + jnt_name + "\" is invalid.");
      return false;
    }
    if (home_pos < min_pos || max_pos < home_pos)
    {
      qt::qErrorBox(this, "Home position of joint \"" + jnt_name + "\" is out of its limit.");
      return false;
    }
  }

  return true;
}

YAML::Node CustomJointsWidget::dump()
{
  YAML::Node node(YAML::NodeType::Sequence);

  for (int row = 0; row < count(); ++row)
  {
    YAML::Node jnt_node(YAML::NodeType::Map);
    jnt_node[kJointNameLabel] = getJointName(row);
    jnt_node[kHomePosLabel] = getHomePosition(row);
    jnt_node[kMinPosLabel] = getMinPosition(row);
    jnt_node[kMaxPosLabel] = getMaxPosition(row);
    jnt_node[kCmdTypeLabel] = getCommandType(row);
    jnt_node[kPGainLabel] = getPGain(row);
    jnt_node[kIGainLabel] = getIGain(row);
    jnt_node[kDGainLabel] = getDGain(row);
    node.push_back(jnt_node);
  }

  return node;
}

void CustomJointsWidget::load(const YAML::Node& node)
{
  for (const auto& jnt_node : node)
  {
    const auto jnt_name = jnt_node[kJointNameLabel].as<QString>();
    const auto row = getRow(jnt_name);
    if (row < 0)
      qt::qErrorBox(this, "\"" + jnt_name + "\" does not exist in the custom joint list.");
    continue;

    setHomePosition(row, jnt_node[kHomePosLabel].as<double>());
    setMinPosition(row, jnt_node[kMinPosLabel].as<double>());
    setMaxPosition(row, jnt_node[kMaxPosLabel].as<double>());
    setCommandType(row, jnt_node[kCmdTypeLabel].as<tobas::joint_control_type_t>());
    setPGain(row, jnt_node[kPGainLabel].as<double>());
    setIGain(row, jnt_node[kIGainLabel].as<double>());
    setDGain(row, jnt_node[kDGainLabel].as<double>());
  }
}

int CustomJointsWidget::count() const
{
  return table_->rowCount();
}

QString CustomJointsWidget::getLinkName(int row) const
{
  const auto cell = qobject_cast<QLabel*>(table_->cellWidget(row, static_cast<int>(LINK_NAME)));
  return cell->text();
}

void CustomJointsWidget::setLinkName(int row, const QString& text)
{
  const auto cell = qobject_cast<QLabel*>(table_->cellWidget(row, static_cast<int>(LINK_NAME)));
  cell->setText(text);
}

QString CustomJointsWidget::getJointName(int row) const
{
  const auto cell = qobject_cast<QLabel*>(table_->cellWidget(row, static_cast<int>(JOINT_NAME)));
  return cell->text();
}

void CustomJointsWidget::setJointName(int row, const QString& text)
{
  const auto cell = qobject_cast<QLabel*>(table_->cellWidget(row, static_cast<int>(JOINT_NAME)));
  cell->setText(text);
}

double CustomJointsWidget::getHomePosition(int row) const
{
  const auto cell = qobject_cast<qt::DoubleSpinBox*>(table_->cellWidget(row, static_cast<int>(HOME_POSITION)));
  return cell->value();
}

void CustomJointsWidget::setHomePosition(int row, double value)
{
  const auto cell = qobject_cast<qt::DoubleSpinBox*>(table_->cellWidget(row, static_cast<int>(HOME_POSITION)));
  cell->setValue(value);
}

double CustomJointsWidget::getMinPosition(int row) const
{
  const auto cell = qobject_cast<qt::DoubleSpinBox*>(table_->cellWidget(row, static_cast<int>(MIN_POSITION)));
  return cell->value();
}

void CustomJointsWidget::setMinPosition(int row, double value)
{
  const auto cell = qobject_cast<qt::DoubleSpinBox*>(table_->cellWidget(row, static_cast<int>(MIN_POSITION)));
  cell->setValue(value);
}

double CustomJointsWidget::getMaxPosition(int row) const
{
  const auto cell = qobject_cast<qt::DoubleSpinBox*>(table_->cellWidget(row, static_cast<int>(MAX_POSITION)));
  return cell->value();
}

void CustomJointsWidget::setMaxPosition(int row, double value)
{
  const auto cell = qobject_cast<qt::DoubleSpinBox*>(table_->cellWidget(row, static_cast<int>(MAX_POSITION)));
  cell->setValue(value);
}

tobas::joint_control_type_t CustomJointsWidget::getCommandType(int row) const
{
  const auto cell = qobject_cast<qt::ComboBox*>(table_->cellWidget(row, static_cast<int>(COMMAND_TYPE)));
  const auto text = cell->currentText();

  if (text == kPositionLabel)
    return tobas::joint_control_type_t::POSITION_CONTROL;
  else if (text == kVelocityLabel)
    return tobas::joint_control_type_t::VELOCITY_CONTROL;
  else if (text == kEffortLabel)
    return tobas::joint_control_type_t::EFFORT_CONTROL;
  else
    throw;
}

void CustomJointsWidget::setCommandType(int row, const tobas::joint_control_type_t& value)
{
  QString text;
  switch (value)
  {
    case tobas::joint_control_type_t::POSITION_CONTROL:
      text = kPositionLabel;
      break;
    case tobas::joint_control_type_t::VELOCITY_CONTROL:
      text = kVelocityLabel;
      break;
    case tobas::joint_control_type_t::EFFORT_CONTROL:
      text = kEffortLabel;
      break;
    default:
      throw;
  }

  const auto cell = qobject_cast<qt::ComboBox*>(table_->cellWidget(row, static_cast<int>(COMMAND_TYPE)));
  cell->setCurrentText(text);
}

double CustomJointsWidget::getPGain(int row) const
{
  const auto cell = qobject_cast<qt::DoubleSpinBox*>(table_->cellWidget(row, static_cast<int>(P_GAIN)));
  return cell->value();
}

void CustomJointsWidget::setPGain(int row, double value)
{
  const auto cell = qobject_cast<qt::DoubleSpinBox*>(table_->cellWidget(row, static_cast<int>(P_GAIN)));
  cell->setValue(value);
}

double CustomJointsWidget::getIGain(int row) const
{
  const auto cell = qobject_cast<qt::DoubleSpinBox*>(table_->cellWidget(row, static_cast<int>(I_GAIN)));
  return cell->value();
}

void CustomJointsWidget::setIGain(int row, double value)
{
  const auto cell = qobject_cast<qt::DoubleSpinBox*>(table_->cellWidget(row, static_cast<int>(I_GAIN)));
  cell->setValue(value);
}

double CustomJointsWidget::getDGain(int row) const
{
  const auto cell = qobject_cast<qt::DoubleSpinBox*>(table_->cellWidget(row, static_cast<int>(D_GAIN)));
  return cell->value();
}

void CustomJointsWidget::setDGain(int row, double value)
{
  const auto cell = qobject_cast<qt::DoubleSpinBox*>(table_->cellWidget(row, static_cast<int>(D_GAIN)));
  cell->setValue(value);
}

QStringList CustomJointsWidget::getLinkNames() const
{
  QStringList res;
  for (int row = 0; row < count(); ++row)
    res.append(getLinkName(row));
  return res;
}

QStringList CustomJointsWidget::getJointNames() const
{
  QStringList res;
  for (int row = 0; row < count(); ++row)
    res.append(getJointName(row));
  return res;
}

QString CustomJointsWidget::getControllerType(int row) const
{
  QString group, controller;

  const auto jnt_name = getJointName(row);
  const auto hi = robot_.hardwareInterface(jnt_name.toStdString());
  switch (hi)
  {
    case hw_interface::POSITION:
      group = "position_controllers";
      break;
    case hw_interface::VELOCITY:
      group = "velocity_controllers";
      break;
    case hw_interface::EFFORT:
      group = "effort_controllers";
      break;
    default:
      throw;
  }

  const auto cmd_type = getCommandType(row);
  switch (cmd_type)
  {
    case tobas::joint_control_type_t::POSITION_CONTROL:
      controller = "JointGroupPositionController";
      break;
    case tobas::joint_control_type_t::VELOCITY_CONTROL:
      controller = "JointGroupVelocityController";
      break;
    case tobas::joint_control_type_t::EFFORT_CONTROL:
      controller = "JointGroupEffortController";
      break;
    default:
      throw;
  }

  return group + "/" + controller;
}

bool CustomJointsWidget::pidEnabled(int row) const
{
  const auto jnt_name = getJointName(row);
  const auto hi = robot_.hardwareInterface(jnt_name.toStdString());
  const auto cmd_type = getCommandType(row);

  if (hi == hw_interface::POSITION && cmd_type == tobas::joint_control_type_t::POSITION_CONTROL)
    return false;
  if (hi == hw_interface::VELOCITY && cmd_type == tobas::joint_control_type_t::VELOCITY_CONTROL)
    return false;
  if (hi == hw_interface::EFFORT && cmd_type == tobas::joint_control_type_t::EFFORT_CONTROL)
    return false;

  return true;
}

int CustomJointsWidget::getRow(const QString& jnt_name)
{
  for (int row = 0; row < count(); ++row)
    if (getJointName(row) == jnt_name)
      return row;

  return -1;
}
}  // namespace setup_assistant
}  // namespace gui
